File: backend/app/schemas/share.py
"""Schemas for public cell share links.

Two sides:
  * ``ShareLinkRead`` — what the OWNER sees in the app after sharing (the token
    to build the URL from, plus expiry/revocation state).
  * ``SharedCellRead`` — what an ANONYMOUS visitor gets from the public
    endpoint. Deliberately minimal: the cell's content plus just enough context
    to label the page. The table's name is NOT exposed — internal naming is
    nobody's business on a public URL.
"""
from datetime import datetime

from pydantic import BaseModel, ConfigDict


class ShareLinkRead(BaseModel):
    """An owner-side view of a share link. The frontend builds the public URL
    as ``/share/{token}``."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    token: str
    row_id: int
    column_id: int
    created_at: datetime
    expires_at: datetime
    revoked_at: datetime | None = None


class SharedCellRead(BaseModel):
    """The public payload. ``content`` is the cell's CURRENT value (these links
    are live), rendered client-side inside a sandboxed iframe."""

    content: str
    # Column label + 1-based row number, so the page can title itself without
    # leaking the table name.
    column_name: str
    row_number: int
    expires_at: datetime

File: frontend/lib/share.ts
import { api } from "./api";

// Mirrors backend app/schemas/share.py.

/** Owner-side view of a public share link. Build the URL with `shareUrl`. */
export interface ShareLink {
  id: number;
  token: string;
  row_id: number;
  column_id: number;
  created_at: string;
  expires_at: string;
  revoked_at: string | null;
}

/** What an anonymous visitor gets. Deliberately minimal — no table name. */
export interface SharedCell {
  content: string;
  column_name: string;
  row_number: number;
  expires_at: string;
}

/** The cell's active link, or null when it isn't shared. */
export function getCellShareLink(
  tableId: number,
  rowId: number,
  columnId: number,
): Promise<ShareLink | null> {
  return api<ShareLink | null>(
    `/library/tables/${tableId}/cells/${rowId}/${columnId}/share`,
  );
}

/** Mint (or re-use) a public link to this cell. Idempotent while one is live. */
export function createCellShareLink(
  tableId: number,
  rowId: number,
  columnId: number,
): Promise<ShareLink> {
  return api<ShareLink>(
    `/library/tables/${tableId}/cells/${rowId}/${columnId}/share`,
    { method: "POST" },
  );
}

/** Kill a link immediately — the public URL 404s from here on. */
export function revokeCellShareLink(linkId: number): Promise<void> {
  return api<void>(`/library/share-links/${linkId}`, { method: "DELETE" });
}

/** Public read — no auth header (the recipient has no account). */
export function getSharedCell(token: string): Promise<SharedCell> {
  return api<SharedCell>(`/share/${encodeURIComponent(token)}`, {
    noAuth: true,
  });
}

/** The absolute URL to hand to someone. */
export function shareUrl(token: string): string {
  const origin =
    typeof window !== "undefined" ? window.location.origin : "";
  return `${origin}/share/${token}`;
}
